plugins/mistralai_sut/modelgauge/suts/vertexai_client.py

```diff
@@ -0,0 +1,83 @@
+import os
+
+import google.auth
+import httpx
+from google.auth.transport.requests import Request
+
+from modelgauge.secret_values import RequiredSecret, SecretDescription
+from pydantic import BaseModel
+
+
+class VertexAIAPIKey(RequiredSecret):
+    """This key is used to request an access token from Google auth, which is then
+    used to call the inference endpoint. It is not used with the endpoint."""
+
+    @classmethod
+    def description(cls) -> SecretDescription:
+        return SecretDescription(
+            scope="vertexai",
+            key="api_key",
+            instructions="See https://cloud.google.com/docs/authentication/api-keys",
+        )
+
+
+class VertexAIProjectId(RequiredSecret):
+    @classmethod
+    def description(cls) -> SecretDescription:
+        return SecretDescription(
+            scope="vertexai",
+            key="project_id",
+            instructions="See https://cloud.google.com/docs/authentication/api-keys",
+        )
+
+
+class VertexAIClient:
+    def __init__(
+        self,
+        publisher: str,
+        model_name: str,
+        model_version: str,
+        streaming: bool,
+        api_key: VertexAIAPIKey,
+        project_id: VertexAIProjectId,
+        region: str = os.environ.get("GOOGLE_REGION"),
+    ):
+        self.publisher = publisher
+        self.model_name = model_name
+        self.model_version = model_version
+        self.api_key = api_key.value
+        self.project_id = project_id.value
+        self.region = region
+        self.streaming = streaming
+
+    def _get_access_token(self) -> str:
+        credentials, _ = google.auth.default(scopes=["https://www.googleapis.com/auth/cloud-platform"])
+        credentials.refresh(Request())
+        return credentials.token
+
+    def _build_endpoint_url(self) -> str:
+        base_url = f"https://{self.region}-aiplatform.googleapis.com/v1/"
+        project_fragment = f"projects/{self.project_id}"
+        location_fragment = f"locations/{self.region}"
+        specifier = "streamRawPredict" if self.streaming else "rawPredict"
+        model_fragment = f"publishers/{self.publisher}/models/{self.model_name}@{self.model_version}"
+        url = f"{base_url}{'/'.join([project_fragment, location_fragment, model_fragment])}:{specifier}"
+        return url
+
+    def _headers(self):
+        headers = {
+            "Authorization": f"Bearer {self._get_access_token()}",
+            "Accept": "application/json",
+        }
+        return headers
+
+    def request(self, req: dict) -> dict:
+        try:
+            client = httpx.Client()
+            response = client.post(self._build_endpoint_url(), json=req, headers=self._headers(), timeout=None)
+            if response.status_code == 200:
+                return response.json()
+            else:
+                raise RuntimeError(f"VertexAI response code {response.status_code}")
```

Comment:
Any production SUT will need to have some retry logic sufficient to overcome transient errors and to keep retrying on rate limits.

Reply:
OK

Reply:
I've added retry logic to the MistralAI sut, and left a comment to revisit this client if/when we use VertexAI again in the future.